src/charm.py

```diff
@@ -3,61 +3,77 @@
 import sys
 sys.path.append('lib')
 
-from ops.charm import CharmBase
-from ops.framework import StoredState
+from ops.charm import CharmBase, CharmEvents
+from ops.framework import EventBase, EventSource, StoredState
 from ops.main import main
 from ops.model import (
     ActiveStatus,
-    BlockedStatus,
-    WaitingStatus,
     MaintenanceStatus,
+    WaitingStatus,
 )
 
-from interfaces import (
-    HTTPInterfaceProvides,
-    MySQLInterfaceRequires,
-)
-from resources import OCIImageResource
+from interface_http import HTTPServer
+from interface_mysql import MySQLServer, MySQLServerError
+from oci_image import OCIImageResource, FetchError
+
+
+class ConfigurePodEvent(EventBase):
+    pass
+
+
+class ServeWebsiteEvent(EventBase):
+    pass
+
+
+class GitLabK8sCharmEvents(CharmEvents):
+    configure_pod = EventSource(ConfigurePodEvent)
+    serve_website = EventSource(ServeWebsiteEvent)
+
+
+class NotLeaderError(Exception):
+    def __init__(self):
+        super().__init__('not leader')
+        self.status = WaitingStatus('Not leader')
```

Comment:
These messages are too vague. Not being a leader is not itself and error. There's something else taking place that doesn't work under those conditions, or some assumption about what it might want to do that cannot be accomplished under those terms. That's what it needs to communicate so that the operator can fix the problem.